### PR TITLE
Metric Rankings

### PR DESCRIPTION
**Pull request by colszowka**
**tl;dr: I want to add annotations to all metrics visible on the site to give a bit of perspective where the given value for the project you're looking at stands in the grander scheme of the ruby ecosystem and need your feedback :heart:**

---

Current state on master:

www ruby-toolbox com_categories_background_jobs 1

After this change:

localhost_5000_categories_background_jobs

Fetching this kind of data is fairly straightforward using postgres percentile functions, like so:

```sql
SELECT unnest(percentile_disc(array[0.1, 0.25, 0.5, 0.75, 0.85, 0.95, 0.99, 1.0])
       WITHIN GROUP (ORDER BY downloads ASC))
   FROM rubygems;
```

While I did some db-backed prototyping I opted to for now go with hard-coded values to first discuss and decide upon some UI/UX headaches this has been giving me.

For now I'm focussing only on the "top row" of project metrics. Later down the road I'd like to expand this:

* Use it for all metrics (including the date ones, i.e. first gem release date etc)
* Add detailed explanation pages upon click on a metric: Showing a distribution graph for the metric, maybe an explanation as to why this metric is interesting when rating a project, and, i.e., the top 100 projects in that measurement

But, here's the current challenges:

## The numbers

To give some background, here's a little breakdown of the distribution of the numbers for the 4 metrics under discussion here:

|      | Gem Downloads | Stargazers | Forks | Watchers | 
|------|---------------|------------|-------|----------| 
| 10%  | 1101          | 0          | 0     | 1        | 
| 25%  | 1926          | 0          | 0     | 1        | 
| 50%  | 3981          | 2          | 0     | 1        | 
| 75%  | 10300         | 7          | 2     | 3        | 
| 85%  | 18807         | 18         | 5     | 6        | 
| 95%  | 79188         | 103        | 21    | 21       | 
| 99%  | 1486655       | 782        | 129   | 76       | 
| 100% | 250153777     | 104573     | 18091 | 6054     | 

As you can see easily, the gap to the absolute top projects is very large, then there is a top 1-10% of projects, and bascially everything else has very similar numbers.

This makes it pretty hard to find meaningful "buckets" to display on the site. I picked the following buckets for display, this is absolutely up for discussion though: `0.5, 0.75, 0.95, 0.99`

Secondly, I find it tough to display this on the site in a clear yet unobtrusive way. I went with arrow-icons and tool-tips, but particularly because of the "top-heavy" buckets I picked, this is somewhat nonsensical:

bildschirmfoto vom 2018-06-26 22 40 00
bildschirmfoto vom 2018-06-26 22 40 07

Maybe a "school grades" ranking system, and/or color coding would work better? With school grades we have the issue that the systems vary vastly across the world. I think color coded indicators might be more clearly distinguishable at a glance, but those are hard to get right with regard to color blind people.

As an additional thought, I don't want to discourage the majority of gem owners whose projects don't "belong to the top 1%" by only showing "negative" indicators. This might be alleviated a bit once other metrics like issue closure rate etc. become part of this, but it's something I'm also trying to pay attention to.

---

So, long story short: I would appreciate your feedback on those two questions:

* Based on the data shown above, what percentile buckets should we pick?
* How can we best display this data in a clear, unobtrusive way within the project view?

Thanks :)

### CODE REVIEW

**Comment by sunito** [comment]
To my eye, a set of colors would look nicer than school grades. Have you thought about displaying a color next to each number? (For example seven colors: red, orange, yellow, lime, green, tuquoise, blue)

**Comment by booch** [comment]
I like the idea of indicating 50th, 75th, 95th, and 99th percentiles. (Maybe change 75th to 80th, though, to have top 1, 5, 20, 50 percentiles, which would be a more even progression.) I wouldn't put any indicator for things below the 50th percentile. I like your icon choice for 50th percentile, especially with the tooltip. It's hard to come up with 3 for the others though. Maybe ▷ for 50th, △ for 75th, ▲ for 95th, and ★ for 99th?

I like the color and battery indicator ideas, but I don't think they map that well to progressively-sized buckets, and I'd prefer to have those progressive buckets.

**Comment by colszowka** [comment]
Thanks for your feedback @sunito and @booch !

The problem with color coding is, apart from aforementioned issue with color-blindness, that I'm not so sure that the spectrum actually is so universal. What color tones would you suggest for the individual buckets?

I also thought whether it might make more sense to actually drop the ranking buckets and just highlight the top X percent with a star.

**Comment by BookOfGreg** [comment]
I feel like the simplest solution would be saying it as text but it does make the page a bit cluttered. Probably chevrons or @booch 's unicode chars may be best.

image

I'm kinda surprised by the stats though, the 99th percentile seems low to me but I suppose you have all the stats, guess the actual distribution is heavily positively skewed, wonder if there's anything that could compensate for that.

File: app/helpers/application_helper.rb [hunk 1]
```diff
@@ -1,14 +1,46 @@
...
+
+  def rank(metric:, value:)
+    return unless metric
+    rank = global_stats.rank metric, value
+    return unless rank
+    content_tag "i", "", class: "fa #{rank_icon(rank)} rank rank-#{rank}", title: rank_tooltip(rank)
+  end
+
+  def rank_icon(rank)
+    case rank
```
Comment (gabrielbaldao) [inline app/helpers/application_helper.rb:24]: What do you think about use
```
return "fa-angle-double-up" if rank == 5

"fa-caret-up"
```

or for future cases

```
rank_icons = { 5: "fa-angle-double-up"}

rank_icons[rank] || "fa-caret-up"
```
